src/main/java/org/skriptlang/skript/lang/condition/DNFConditionalBuilder.java

```diff
@@ -0,0 +1,172 @@
+		return this;
+	}
+
+	/**
+	 * Adds conditionals to the root node via the OR operator: {@code (existing) || newA || newB || ...}.
+	 * If the root is currently AND, a new OR root node is created containing the previous root and the new conditionals.
+	 * @param orConditionals conditionals to OR to the existing conditional.
+	 * @return the builder
+	 */
+	@SafeVarargs
+	@Contract("_ -> this")
+	public final DNFConditionalBuilder<T> or(Conditional<T>... orConditionals) {
+		if (root == null) {
+			root = new CompoundConditional<>(Operator.OR, orConditionals);
+			return this;
+		}
+
+		// unroll conditionals if they're ORs
+		List<Conditional<T>> newConditionals = unroll(List.of(orConditionals), Operator.OR);
+
+		// Since DNF is a series of ANDs, ORed together, we can simply add these to the root if it's already OR.
+		if (root.getOperator() == Operator.OR) {
+			root.addConditionals(newConditionals);
+			return this;
+		}
+		// otherwise we need to nest the AND/NOT condition within a new root with OR operator.
+		newConditionals.add(0, root);
+		root = new CompoundConditional<>(Operator.OR, newConditionals);
+		return this;
+	}
+
+	/**
+	 * Adds a negated conditional to the root node via the AND and NOT operators: {@code (existing) && !new}.
+	 * @param conditional The conditional to negate and add.
+	 * @return the builder
+	 */
+	@Contract("_ -> this")
+	public DNFConditionalBuilder<T> andNot(Conditional<T> conditional) {
+		return and(negate(conditional));
+	}
+
+	/**
+	 * Adds a negated conditional to the root node via the OR and NOT operators: {@code (existing) || !new}.
+	 * @param conditional The conditional to negate and add.
+	 * @return the builder
+	 */
+	@Contract("_ -> this")
+	public DNFConditionalBuilder<T> orNot(Conditional<T> conditional) {
+		return or(negate(conditional));
+	}
+
+	/**
+	 * Adds conditionals to the root node via the AND or OR operators.
+	 * A helper for dynamically adding conditionals.
+	 * @param or Whether to use OR (true) or AND (false)
+	 * @param conditionals The conditional to add.
+	 * @return the builder
+	 */
+	@SafeVarargs
+	@Contract("_,_ -> this")
+	public final DNFConditionalBuilder<T> add(boolean or, Conditional<T>... conditionals) {
+		if (or)
+			return or(conditionals);
+		return and(conditionals);
+	}
+
+	/**
+	 * Unrolls nested conditionals which are superfluous:
+	 * {@code a || (b || c) -> a || b || c}
+	 * @param conditionals A collection of conditionals to unroll.
+	 * @param operator Which operator to unroll.
+	 * @return A new list of conditionals without superfluous nesting.
+	 */
+	@Contract("_,_ -> new")
+	private static <T> List<Conditional<T>> unroll(Collection<Conditional<T>> conditionals, Operator operator) {
+		List<Conditional<T>> newConditionals = new ArrayList<>();
+		for (Conditional<T> conditional : conditionals) {
+			if (conditional instanceof CompoundConditional<T> compound && compound.getOperator() == operator) {
```

Comment:
Hmm it might be a good idea to have better support with this. Maybe conditions could have an unroll method (by default returns itself?). I'm just thinking if there are other implementations that would need special handling like this.

Reply:
I'm not really sure if it'd be valuable as a method for all conditions, since it's pretty context-dependent.
How many layers should it unroll? Should it modify the conditional or return a new conditional or even a list of conditionals? I think if it comes up in the future we can address it but for now I don't think it's worth the investment